### PR TITLE
Addition of the class `FluxMetaData`

### PR DESCRIPTION
**Pull request by bkhelifi**
**Description**

This pull request is associated to the Issue #4790. It contains the basic structure of this class, requiring still some discussion with the reviewers. 
Separating the internal data format of Gammapy with the serialization, the functions `from_dict`, `to_table` are however associated to the serialization. And with these class attributes, we are going beyond the GADF format.

**Dear reviewer**
The implementation has been achieved by making the full integration of this class in the `estimators` sub-package. For the reviewing purpose, I have splitted my work into two:
- this PR with the basic class
- a second one about the usage of the class in `estimators` sub-package (with some impact on the map sub-package because of missing propagation of metadata in some functions). The draft of this work can be found the branch 'Issue4790' in my fork (https://github.com/bkhelifi/gammapy/tree/Issue4790)

### CODE REVIEW

File: gammapy/estimators/metadata.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+from typing import Optional
+import numpy as np
+from scipy import stats
+from astropy.coordinates import SkyCoord
+from pydantic import ValidationError, validator
+from gammapy.utils.metadata import CreatorMetaData, MetaData
+
+__all__ = ["FluxMetaData"]
+
+SEDTYPE = ["dnde", "flux", "eflux", "e2dnde", "likelihood"]
```
Comment (adonath) [inline gammapy/estimators/metadata.py:12]: The better way to handle these is with an `Enum`, or `Literal`

File: gammapy/estimators/metadata.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        table : `~astropy.table.Table`
+            Flux table.
+        """
+
+        if self.sed_type:
+            table.meta["SED_TYPE"] = self.sed_type
+        if self.sed_type_init:
+            table.meta["SEDTYPEI"] = self.sed_type_init
+        if self.ul_conf:
+            table.meta["UL_CONF"] = str(self.ul_conf)
```
Comment (adonath) [inline gammapy/estimators/metadata.py:206]: The type conversion her is not needed, pedantic will guarantee that it's as thing already...

File: gammapy/estimators/metadata.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+from scipy import stats
+from astropy.coordinates import SkyCoord
+from pydantic import ValidationError, validator
+from gammapy.utils.metadata import CreatorMetaData, MetaData
+
+__all__ = ["FluxMetaData"]
+
+SEDTYPE = ["dnde", "flux", "eflux", "e2dnde", "likelihood"]
+FPFORMAT = ["gadf-sed", "lightcurve"]
+STANDARD_KEYS = [
```
Comment (adonath) [inline gammapy/estimators/metadata.py:14]: I would propose to handle this in a more generic way. Introduce a dict that maps the meta data attribute to its FITS name:

```python
FITS_META_KEYS = {
    "sed_type": "SED_TYPE",
    "instrument": "INSTRU",
...
}
```
You can also introduce the inverse mapping and the later use it avoid all the hard coded `if` statements.

File: gammapy/estimators/metadata.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+    "INSTRU",
+]
+
+log = logging.getLogger(__name__)
+
+
+class FluxMetaData(MetaData):
+    """Metadata containing information about the FluxPoints and FluxMaps.
+
+    Parameters
```
Comment (adonath) [inline gammapy/estimators/metadata.py:35]:
```suggestion
    Attributes
```

File: gammapy/estimators/metadata.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+        if v is None:
+            return SkyCoord(np.nan, np.nan, unit="deg", frame="icrs")
+        elif isinstance(v, SkyCoord):
+            return v
+        else:
+            raise ValidationError(
+                f"Incorrect position. Expect SkyCoord got {type(v)} instead."
+            )
+
+    @validator("sed_type")
```
Comment (adonath) [inline gammapy/estimators/metadata.py:98]: If you introduce an `Enum` or `Literal` this can be removed...

File: gammapy/estimators/metadata.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+    @classmethod
+    def from_default(cls):
+        return cls(
+            # creation=CreatorMetaData.from_default(), target_position=None, gti=None
+            creation=CreatorMetaData.from_default()
+        )
+
+    @classmethod
+    def from_dict(cls, data):
```
Comment (adonath) [inline gammapy/estimators/metadata.py:139]: I think this should explicitly mention FITS in the method name.

**Review by adonath — COMMENTED** [review]
Thanks @bkhelifi, I have later some first comments.

File: gammapy/estimators/metadata.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+        return v
+
+    @validator("sed_type_init")
+    def validate_sed_type_init(cls, v):
+        if isinstance(v, str):
+            if v not in SEDTYPE:
+                raise ValidationError(f"Incorrect [sed_type_init]. Expect {SEDTYPE}")
+        return v
+
+    @validator("obs_ids")
```
Comment (registerrier) [inline gammapy/estimators/metadata.py:112]: This does not need a validator as the type checking/casting will do that for you.

File: gammapy/estimators/metadata.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+    sed_type_init: Optional[str]
+    ul_conf: Optional[float]
+    n_sigma: Optional[float]
+    n_sigma_ul: Optional[float]
+    sqrt_ts_threshold_ul: Optional[float]
+    n_sigma_sensitivity: Optional[float]
+    # gti: Optional[GTI]
+    target_name: Optional[str]  # Are these 2 fields really optional?
+    target_position: Optional[SkyCoord]
+    obs_ids: Optional[list[int]]
```
Comment (registerrier) [inline gammapy/estimators/metadata.py:81]:
```suggestion
    obs_ids: Optional[List[int]]
```

File: gammapy/estimators/metadata.py [hunk 9]
```diff
@@ -0,0 +1,256 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+import logging
+from typing import Optional
```
Comment (registerrier) [inline gammapy/estimators/metadata.py:3]:
```suggestion
from typing import Optional, List
```

File: gammapy/estimators/metadata.py [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    ul_conf: Optional[float]
+    n_sigma: Optional[float]
+    n_sigma_ul: Optional[float]
+    sqrt_ts_threshold_ul: Optional[float]
+    n_sigma_sensitivity: Optional[float]
+    # gti: Optional[GTI]
+    target_name: Optional[str]  # Are these 2 fields really optional?
+    target_position: Optional[SkyCoord]
+    obs_ids: Optional[list[int]]
+    dataset_names: Optional[list[str]]  # Are these 2 fields really optional?
```
Comment (registerrier) [inline gammapy/estimators/metadata.py:82]:
```suggestion
    dataset_names: Optional[List[str]]  # Are these 2 fields really optional?
```

Comment (registerrier) [reply gammapy/estimators/metadata.py:82]: If they are not optional you won't be able to read external or previously generated flux points.

File: gammapy/estimators/metadata.py [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+class FluxMetaData(MetaData):
+    """Metadata containing information about the FluxPoints and FluxMaps.
+
+    Parameters
+    ----------
+    sed_type : {"dnde", "flux", "eflux", "e2dnde", "likelihood"}, optional
+        SED type.
+    sed_type_init : {"dnde", "flux", "eflux", "e2dnde", "likelihood"}, optional
+        SED type of the initial data.
+    ul_conf : float, optional
```
Comment (registerrier) [inline gammapy/estimators/metadata.py:41]: isn't this repeating `n_sigma_ul`?

**Review by registerrier — COMMENTED** [review]
Thanks @bkhelifi ! I have left some first comments.

File: gammapy/estimators/metadata.py (hunk 3 above)
Comment (bkhelifi) [reply gammapy/estimators/metadata.py:14]: I like it a lot. Thanks for the proposal... knowing that we are already at the edge with serialisation.

File: gammapy/estimators/metadata.py (hunk 2 above)
Comment (bkhelifi) [reply gammapy/estimators/metadata.py:206]: It is not in the wrong direction? Here, we fill an astropy table, not a pedantic object. I think that the casting is needed

File: gammapy/estimators/metadata.py (hunk 1 above)
Comment (adonath) [reply gammapy/estimators/metadata.py:12]: Here is an example how we use `Enum` in other places: https://github.com/gammapy/gammapy/blob/main/gammapy/analysis/config.py#L69

File: gammapy/estimators/metadata.py [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+                raise ValidationError(f"Incorrect [sed_type_init]. Expect {SEDTYPE}")
+        return v
+
+    @validator("obs_ids")
+    def validate_obs_ids(cls, v):
+        if isinstance(v, list) and not all(np.isfinite(_) for _ in v):
+            raise ValidationError("Incorrect [obs_ids]. Expect [int]")
+        return v
+
+    @validator("creation")
```
[the remark this answers is not in the document]
Comment (adonath) [reply gammapy/estimators/metadata.py:118]: Yes, no validator is needed here.

File: gammapy/estimators/metadata.py [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+        Threshold on the square root of the likelihood value above which upper limits should be used.
+    n_sigma_sensitivity : float, optional
+        Sigma number for which the flux sensitivity is computed
+    # gti : `~gammapy.data.gti`, optional
+    #     used Good Time Intervals.
+    target_name : str, optional
+        Name of the target.
+    target_position : `~astropy.coordinates.SkyCoord`, optional
+        Coordinates of the target.
+    obs_ids : list of int, optional
```
[the remark this answers is not in the document]
Comment (bkhelifi) [reply gammapy/estimators/metadata.py:57]: I have no idea! Which type is it on the observation meta class? An int?

**Comment by adonath** [comment]
@bkhelifi Please rebase and fix merge conflicts.

**Comment by registerrier** [comment]
Hi @bkhelifi . Can you rebase? 
I think you can keep the format argument and check in this PR and we can see in a different PR how we can deal with the format specifically.

**Comment by bkhelifi** [comment]
In this PR, I have kept the test on the format value in the main to_header function of utils.metadata.
In a separate PR, I will propose to remove it...

File: gammapy/data/metadata.py [hunk 14]
```diff
@@ -80,7 +80,16 @@ class ObservationMetaData(MetaData):
...
+        """Create and fill the observation metadata from the event list metadata.
+
+        Parameters
+        ----------
+        header : dict
+            Input FITS header.
+        format : str
+            The header data format. Default is gadf.
+        """
+        meta = super(ObservationMetaData, cls).from_header(header, format)
```
Comment (registerrier) [inline gammapy/data/metadata.py:92]:
```suggestion
        meta = super().from_header(header, format)
```

File: gammapy/data/tests/test_metadata.py [hunk 15]
```diff
@@ -79,8 +79,9 @@ def test_observation_metadata_from_header(hess_eventlist_header):
 
 @requires_data()
 def test_observation_metadata_bad(hess_eventlist_header):
-    with pytest.raises(ValueError):
-        ObservationMetaData.from_header(hess_eventlist_header, format="bad")
+    # The test on the format has been commented in the code
+    # with pytest.raises(ValueError):
+    #     ObservationMetaData.from_header(hess_eventlist_header, format="bad")
```
Comment (registerrier) [inline gammapy/data/tests/test_metadata.py:84]:
```suggestion
    # TODO: adapt with proper format handling
    with pytest.raises(ValueError):
        ObservationMetaData.from_header(hess_eventlist_header, format="bad")
```

File: gammapy/estimators/metadata.py [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+    _tag: ClassVar[Literal["flux"]] = "flux"
+    sed_type: Optional[SEDTYPEEnum] = None
+    sed_type_init: Optional[SEDTYPEEnum] = None
+    n_sigma: Optional[float] = None
+    n_sigma_ul: Optional[float] = None
+    sqrt_ts_threshold_ul: Optional[float] = None
+    n_sigma_sensitivity: Optional[float] = None
+    target: Optional[TargetMetaData] = None
+    # ToDo: add obs_ids: Optional[List[int]] and instrument: Optional[str], or a List[ObsInfoMetaData]
+    # ToDo : add dataset_names: Optional[List[str]]
```
Comment (registerrier) [inline gammapy/estimators/metadata.py:75]:
```suggestion
    # TODO: add obs_ids: Optional[List[int]] and instrument: Optional[str], or a List[ObsInfoMetaData]
    # TODO : add dataset_names: Optional[List[str]]
```

**Review by registerrier — APPROVED** [review]
Thanks @bkhelifi . I have left minor comments inline

**Pull request merged**